### DISCUSSION TITLE
How do I change the DB password in encrypted drift switch to the new password

### DISCUSSION THREAD

**Discussion by sharathpanily** [simolus3/drift#2327] [Q&A]
In order to change the DB key when using encrypted drift - it looks like I can use the following code (assuming the last schemaVersion of my DB was 4), where Env.pDbKey contains the new DbKey. I've verified that this works.

```
 @override
  MigrationStrategy get migration {
    return MigrationStrategy(
      onCreate: (Migrator m) async {
        await m.createAll();
      },
      onUpgrade: (Migrator m, int from, int to) async {
        try {
          if (from < 5) {
            await customStatement('PRAGMA rekey = ${Env.pDbKey}');
          }
        } catch (error) {
          dbLogger.e(error);
        }
      },
    );
  }
```
What I'm unsure about is how/when to switch to the new password - from this discussion (https://github.com/simolus3/drift/discussions/1402) it looks like the DB has to be opened/connected to before switching to the new key. Implying that the sequence of operations should to be - 

1. Open DB with oldDbKey
2. Re-key as shown above
3. Then on the next session - detect that the DB needs to use the new key - either through the oldDbKey throwing an exception or by detecting the `schemaVersion`

What I'm currently missing is the code to carry out step 3 - not sure which exception I can catch or how to get the `schemaVersion` while opening the connection during LazyDatabase initialization. I've tried the code below without success. Any help would be much appreciated.
```
LazyDatabase _openConnection() {
  final String oldDbKey = Env.oldDbKey;
  final String newDbKey = Env.pDbKey;

  return LazyDatabase(() async {
    final dbFolder = await getApplicationDocumentsDirectory();

    // encrypted drift DB
    final dbPath = p.join(dbFolder.path, 'myappdb.sqlite');
    try {
      return EncryptedExecutor(path: dbPath, password: oldDbKey);
    } catch (e) {
      dbLogger.e(e);
      return EncryptedExecutor(path: dbPath, password: newDbKey);
    }
  });
}
```

**Comment by simolus3** [comment]
The [documentation](https://www.zetetic.net/sqlcipher/sqlcipher-api/#rekey) for `PRAGMA rekey` says:

> It can be called at any time once the database is readable.

So it seems like you won't need a migration at all and you can just issue the `rekey` pragma while the database is open. Doing it at the beginning (for instance in a migration) might be a good idea though, I imagine the `rekey` pragma is pretty expensive for large databases and might cause a delay to other operations.

Before calling the migration strategy, drift needs to open the database. Otherwise, we have no way of knowing which schema version is currently embedded in the database. For security reasons, the schema version is encrypted as well though. So if you want to try multiple keys and pick the wrong key first, drift won't be able to open the database at all and all statements will fail without even getting to the point where your `onUpgrade` callback could be called.

I think you can put the key adaption logic in your `LazyDatabase` by something like this:

```dart
import 'package:sqflite_sqlcipher/sqflite.dart';
// ....

  return LazyDatabase(() async {
    final dbFolder = await getApplicationDocumentsDirectory();

    // encrypted drift DB
    final dbPath = p.join(dbFolder.path, 'myappdb.sqlite');
    final rawDb = await openDatabase(dbPath, password: oldDbKey);

    // see if the database works
    try {
      await rawDb.execute('pragma user_version');
     // it does, so it's on the old key
      await rawDb.execute("pragma rekey = '$newDbKey'"); // please be careful here to not introduce an SQL injection vulnerability
    } on DatabaseException catch (e) {
      // The old key is probably invalid, so we're using the new db key already.
    } finally {
      await rawDb.close();
    }

    // At this point, the database is migrated to the new key
    return EncryptedExecutor(path: dbPath, password: newDbKey);
  });
```

**Reply by sharathpanily** [reply]
So I tried the code above - I cannot find either `openDatabase` or a `DatabaseException` in drift - what am I missing? Is there a newer version of Drift I should try/use that has this functionality. Currently using the following:

```
drift: ^2.5.0
encrypted_drift:
  git:     
    url: https://github.com/simolus3/drift.git
    path: extras/encryption

dev_dependencies:
  drift_dev: ^2.5.0
```

Also, I'm testing another workaround right now using `FlutterSecureStorage` to keep track of schemaVersion.

**Reply by simolus3** [reply]
> I cannot find either `openDatabase` or a `DatabaseException` in drift

They're using the `sqflite_sqlcipher` package, just like `encrypted_drift`.

If you don't have a strong reason to use the `encrypted_drift` package, I recommend a `NativeDatabase` with `sqlcipher_flutter_libs`, which uses `dart:ffi` instead of (more expensive) platform channels. That variant is [documented here](https://drift.simonbinder.eu/docs/other-engines/encryption/#encrypted-version-of-a-nativedatabase), it also makes it much easier to implement the dynamic key change

```dart
NativeDatabase(
  File(...),
  setup: (rawDb) {
    assert(database.select('PRAGMA cipher_version;').isNotEmpty); // Make sure we have sqlcipher

    rawDb.execute("PRAGMA key = 'oldKey';");
    try {
      rawDb.execute('pragma user_version');
     // it does, so it's on the old key
      rawDb.execute("pragma rekey = '$newDbKey'"); // please be careful here to not introduce an SQL injection vulnerability
    } on SqliteException catch (e) {
      // The old key is probably invalid, so we're using the new db key already.
    }
  }
);
```

**Reply by sharathpanily** [reply]
@simolus3 - I've been trying to use the code you suggested and variations on it - including the sqflite_sqlcipher helps get rid of those errors and makes those functions usable. That said - the error I'm running into now is that when running `await rawDb.execute("pragma rekey = '$newDbKey'");` - it seems like flutter goes directly to `return EncryptedExecutor(path: dbPath, password: newDbKey);` despite the rekey operation not being complete.

If I had to guess - its because the `await rawDb.execute` doesn't actually involve waiting as its return type is a `Future<void>`

I'm still trying another approach and I had a related question - is there a way to detect that the DB rekey is done or that a DB migration is done (if the rekey is done using a migration)

**Reply by simolus3** [reply]
>its because the `await rawDb.execute` doesn't actually involve waiting as its return type is a `Future<void>`

But it should still await the operation.

> is there a way to detect that the DB rekey is done

It should be done after you await the statement (no await necessary with `NativeDatabase.setup` which is synchronous). For migrations, the migration is done when the first statement runs.
